javaslang/src/main/java/javaslang/collection/HashSet.java

```diff
@@ -473,7 +473,7 @@ private HashSet(HashArrayMappedTrie<T, T> tree) {
 
     @Override
     public HashSet<T> add(T element) {
-        return new HashSet<>(tree.put(element, element));
+        return contains(element) ? this : new HashSet<>(tree.put(element, element));
```

Comment:
If key is present, tree.put(k,v) needs to return a new instance for Maps and the same instance for Sets. Therefore we currently perform an additional 'contains' check for Sets. A future optimization may add an additional flag `replace` to the backing HAMT.put() / RedBlackTree.insert() methods. Sets set it to replace=false, Maps set it to replace=true. But for now this solution is working for us.